### DISCUSSION TITLE
How to apply slider to a gltf object and a THREE object (cylinder for example) ?

### DISCUSSION THREAD

**Discussion by J12My** [enable3d/enable3d#274] [Help / Questions]
Hello,

I can't add a slider constraint between a cylinder and a imported gltf object, could you help me ?

<img width="988" alt="Capture d’écran 2022-05-15 à 13 36 09" src="https://user-images.githubusercontent.com/70979321/168470791-66a515bb-8b39-453d-ab1b-104c6e37cb86.png">

it's look like simple but I'm completely new in programmation I can't find the solution alone.

the yellow cylinder is fix and the idea is to have the grey parts with a sliding constraint in x, z

this is the code I use

`
const { AmmoPhysics, PhysicsLoader } = ENABLE3D

const MainScene = () => {
  const scene = new THREE.Scene()
  scene.background = new THREE.Color(0x000000)

  const camera = new THREE.PerspectiveCamera(50, window.innerWidth / window.innerHeight, 0.1, 6000)
  camera.position.set(1200, 1200, 2400)

  const renderer = new THREE.WebGLRenderer()
  renderer.setSize(window.innerWidth, window.innerHeight)
  document.body.appendChild(renderer.domElement)

  const controls = new THREE.OrbitControls(camera, renderer.domElement)

  scene.add(new THREE.HemisphereLight(0xffffbb, 0x080820, 1))
  scene.add(new THREE.AmbientLight(0x666666))
  const light = new THREE.DirectionalLight(0xdfebff, 1)
  light.position.set(2000, 2000, 2000)
  light.position.multiplyScalar(1.3)

  // add axis
  var axes = new THREE.AxesHelper(800);
  scene.add(axes);

  // initialize physics
  const physics = new AmmoPhysics(scene)
  physics.debug.enable(true)

  // add a ground
  physics.add.ground({ width: 3000, height: 3000 })

  //add pipe with physics
  var pipeGeometry = new THREE.CylinderGeometry(10, 10, 200, 200)
  var pipeMaterial = new THREE.MeshToonMaterial({color: 0xFFCC00})
  const pipe = new THREE.Mesh(pipeGeometry, pipeMaterial)
  pipe.position.set(0, 100, 0)
  pipe.castShadow = true;
  pipe.receiveShadow = true;
  scene.add(pipe)
  physics.add.existing(pipe, { mass: 0 })
  pipe.body.setCollisionFlags(1) // make it kinematic
  pipe.body.setFriction(0)

  // add pads with physics
  var loader = new THREE.GLTFLoader()

  loader.load('/enable3d/enable3d-Blank/src/assets/glb/pad01.glb', function (gltf) {
    const padX1 = gltf.scene.children[0]
    padX1.position.setY(0)
    padX1.position.setZ(200)

    padX1.traverse(child => {
      if (child.isMesh) {
        child.castShadow = child.receiveShadow = true
        child.material.metalness = 0.5
        child.material.roughness = 0.2
      }
    })

    scene.add(padX1)
    physics.add.existing(padX1, { shape: 'convex' })
    padX1.body.setFriction(0)        
  })
/*
  physics.add.constraints.slider(pipe.body, padX1.body, {
    frameA: { x: 0, y: 0, z: 0 },
    frameB: { x: 0, y: 0, z: 0 },
    linearLowerLimit: 0,
    linearUpperLimit: 300
    //console.log("constraints")
  })
*/

  loader.load('/enable3d/enable3d-Blank/src/assets/glb/pad01.glb', function (gltf) {
    const padX2 = gltf.scene.children[0]
    padX2.position.setY(0)
    padX2.position.setZ(-200)
    padX2.rotation.y = Math.PI

    padX2.traverse(child => {
      if (child.isMesh) {
        child.castShadow = child.receiveShadow = true
        child.material.metalness = 0.5
        child.material.roughness = 0.2
      }
    })

    scene.add(padX2)
    physics.add.existing(padX2, { shape: 'convex' })
    padX2.body.setFriction(0)
  })

  loader.load('/enable3d/enable3d-Blank/src/assets/glb/pad01.glb', function (gltf) {
    const padY1 = gltf.scene.children[0]
    padY1.position.setY(0)
    padY1.position.setX(200)
    padY1.rotation.y = 0.5 * Math.PI

    padY1.traverse(child => {
      if (child.isMesh) {
        child.castShadow = child.receiveShadow = true
        child.material.metalness = 0.5
        child.material.roughness = 0.2
      }
    })

    scene.add(padY1)
    physics.add.existing(padY1, { shape: 'convex' })
    padY1.body.setFriction(0)
  })

  loader.load('/enable3d/enable3d-Blank/src/assets/glb/pad01.glb', function (gltf) {
    const padY2 = gltf.scene.children[0]
    padY2.position.setY(0)
    padY2.position.setX(-200)
    padY2.rotation.y = -0.5 * Math.PI

    padY2.traverse(child => {
      if (child.isMesh) {
        child.castShadow = child.receiveShadow = true
        child.material.metalness = 0.5
        child.material.roughness = 0.2
      }
    })

    scene.add(padY2)
    physics.add.existing(padY2, { shape: 'convex' })
    padY2.body.setFriction(0)
  })

  const clock = new THREE.Clock()

  const animate = () => {
    //pipe.body.needUpdate = true // this is how you update kinematic bodies
    //physics.destroy(pipe)
    //pipe.scale.x += 0.05;
    //pipe.scale.z += 0.05;
    //physics.add.existing(pipe, { mass: 0 })

    // update physics
    physics.update(clock.getDelta() * 1000)
    // update the physics debugger
    physics.updateDebugger()

    renderer.render(scene, camera)
    requestAnimationFrame(animate)
  }
  requestAnimationFrame(animate)
}
PhysicsLoader('/enable3d/enable3d-Blank/src/lib/ammo/kripken', () => MainScene())

console.log(`three.js version "${THREE.REVISION}"`)

`

when I tried to use this

`
physics.add.constraints.slider(pipe.body, padX1.body, {
    frameA: { x: 0, y: 0, z: 0 },
    frameB: { x: 0, y: 0, z: 0 },
    linearLowerLimit: 0,
    linearUpperLimit: 300
    //console.log("constraints")
  })
`

I have an error because padX1 is undefined

and when I tried to insert this part of code into the loader, like that :

`
loader.load('/enable3d/enable3d-Blank/src/assets/glb/pad01.glb', function (gltf) {
    const padX1 = gltf.scene.children[0]
    padX1.position.setY(0)
    padX1.position.setZ(200)

    padX1.traverse(child => {
      if (child.isMesh) {
        child.castShadow = child.receiveShadow = true
        child.material.metalness = 0.5
        child.material.roughness = 0.2
      }
    })

    scene.add(padX1)
    physics.add.existing(padX1, { shape: 'convex' })
    padX1.body.setFriction(0)        
  })

 physics.add.constraints.slider(pipe.body, padX1.body, {
    frameA: { x: 0, y: 0, z: 0 },
    frameB: { x: 0, y: 0, z: 0 },
    linearLowerLimit: 0,
    linearUpperLimit: 300
    //console.log("constraints")
  })
`

My web browser have an error about memory and the renderer is stuck

do you have an exemple about how to use a slider constraint with an .glb object ?

thanks in advance for your help

Jimmy

**Comment by vitalspace** [comment]
I don't understand exactly what you want to do, but in any case if you want to move an object in any of its axes you have to update its body.

Maybe this can help you:

https://github.com/enable3d/enable3d-website/blob/master/src/examples/teleport-a-dynamic-body.html

Also: 

Update (delta, time){
this.object.body.needUpdate = true;
this.object.rotation.z += 0.05;
this.object.position.x + 0.5;
}

**Reply by J12My** [reply]
Hi thank you for your answer, to be able to move an object (grey in the picture), in animate(), I had to destroy the physic object, apply the new position, then apply the physic...
also I notice that there is a small gap between the objects (grey and yellow) when the collision is detected.

but the main problem was that, I can't apply a slider constraint to lock the grey objects movements only in, or x, or z translation.
to add the slider constraint I created an ExtendedOdject3d
